File: Xamarin/Xamarin.Native/Xamarin.Native.iOS/ViewControllers/PerformanceTestController.cs
using CoreFoundation;
using Foundation;
using System;
using System.Diagnostics;
using System.Drawing;
using System.Threading.Tasks;
using UIKit;
using Xamarin.Services;

namespace Xamarin.Native.iOS.ViewControllers
{
    public partial class PerformanceTestController : UIViewController
    {
        Stopwatch stopwatch;

        public PerformanceTestController(IntPtr handle) : base(handle)
        {
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            PerformanceTestService.Instance.PiCalculationCompleted += PiCalculationCompleted;
        }

  		partial void StartCalculation (UIButton sender)
		{
			resultView.Text = @"";
			RefreshUI(true);
			stopwatch = new Stopwatch();
			stopwatch.Start();

			var digits = Convert.ToInt32(digitsEntry.Text);
			DispatchQueue.DefaultGlobalQueue.DispatchAsync(() =>
			{
				PerformanceTestService.Instance.CalculatePi(digits);
			});
		}

		private void PiCalculationCompleted(string result)
		{
			stopwatch.Stop();

			DispatchQueue.MainQueue.DispatchAsync(() =>
			{
				resultView.Text = result;
				timeLabel.Text = stopwatch.GetDurationInSeconds();
				RefreshUI(false);
			});
		}

		private void RefreshUI(bool isCalculating)
		{
			startButton.Hidden = isCalculating;
			digitsEntry.Enabled = isCalculating;
			activityIndicator.Hidden = !isCalculating;
		}


    }
}

File: Native.iOS/Native.iOS/ViewControllers/PerformanceTestController.m
//
//  PerformanceTestController.m
//  Native.iOS
//
//  Created by Pawel Grzmil on 17/01/16.
//  Copyright © 2016 Pawel Grzmil. All rights reserved.
//

#import "PerformanceTestController.h"
#import "PerformanceTestService.h"
#import "Stopwatch.h"

@interface PerformanceTestController ()
@property (weak, nonatomic) IBOutlet UIButton *startButton;
@property (weak, nonatomic) IBOutlet UITextField *digitsEntry;
@property (weak, nonatomic) IBOutlet UILabel *timeLabel;
@property (weak, nonatomic) IBOutlet UITextView *resultView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (strong, nonatomic) Stopwatch* stopwatch;

@end
@implementation PerformanceTestController

-(void)viewDidLoad{
    [super viewDidLoad];
    self.title = @"Test obliczeń";
    self.stopwatch = [Stopwatch new];
}

- (IBAction)startCalculation:(id)sender {
    [self.stopwatch start];
    [self refreshUI:true];
    
    int digits = [self.digitsEntry.text intValue];
    
    dispatch_async( dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString* pi = [[PerformanceTestService instance] calculatePi:digits];
        [self.stopwatch stop];
        dispatch_async( dispatch_get_main_queue(), ^{
            self.resultView.text = pi;
            self.timeLabel.text = [self.stopwatch getDurationInSeconds];
            [self refreshUI:false];
        });
    });
    
}

- (void)refreshUI: (BOOL)isDownloading{
    self.activityIndicator.hidden = !isDownloading;
    self.startButton.hidden = isDownloading;
    self.digitsEntry.userInteractionEnabled = !isDownloading;
}

@end
